File: backend/app/services/auth_service.py
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.core.security import create_access_token, encrypt_text
from app.db.models.user import User
from app.integrations.github import exchange_code_for_token, get_user


async def github_oauth_callback(code: str, db: Session) -> str:
    access_token = await exchange_code_for_token(code)
    github_user = await get_user(access_token)

    github_id = str(github_user["id"])
    username = github_user.get("login", "unknown")

    stmt = select(User).where(User.github_id == github_id)
    user = db.scalar(stmt)

    if user is None:
        user = User(
            github_id=github_id,
            username=username,
            email=github_user.get("email"),
            avatar_url=github_user.get("avatar_url"),
            access_token=encrypt_text(access_token),
        )
        db.add(user)
    else:
        user.username = username
        user.email = github_user.get("email")
        user.avatar_url = github_user.get("avatar_url")
        user.access_token = encrypt_text(access_token)

    db.commit()
    db.refresh(user)

    return create_access_token(str(user.id))

File: backend /src/modules/auth/auth.service.ts
import jwt from "jsonwebtoken";
import { prisma } from "../../config/db";
import { env } from "../../config/env";
import { githubService } from "../../integrations/github";

export const authService = {
  getGithubLoginUrl: () => {
    const params = new URLSearchParams({
      client_id: env.GITHUB_CLIENT_ID,
      redirect_uri: env.GITHUB_REDIRECT_URI,
      scope: "repo read:user user:email workflow"
    });

    return `https://github.com/login/oauth/authorize?${params.toString()}`;
  },

  handleGithubCallback: async (code: string) => {
    const accessToken = await githubService.exchangeCodeForToken(code);
    const githubUser = await githubService.getUser(accessToken);

    const user = await prisma.user.upsert({
      where: { githubId: String(githubUser.id) },
      update: {
        username: githubUser.login,
        email: githubUser.email ?? `${githubUser.login}@users.noreply.github.com`,
        avatarUrl: githubUser.avatar_url,
        accessToken
      },
      create: {
        githubId: String(githubUser.id),
        username: githubUser.login,
        email: githubUser.email ?? `${githubUser.login}@users.noreply.github.com`,
        avatarUrl: githubUser.avatar_url,
        accessToken
      }
    });

    const token = jwt.sign(
      {
        userId: user.id,
        githubId: user.githubId,
        username: user.username
      },
      env.JWT_SECRET,
      { expiresIn: "7d" }
    );

    return {
      token,
      user: {
        id: user.id,
        username: user.username,
        email: user.email,
        avatarUrl: user.avatarUrl
      }
    };
  }
};
